modules/REST/V1/Languages.php

```diff
@@ -0,0 +1,865 @@
+		$prepared = (array) $this->prepare_item_for_database( $request );
+
+		if ( ! empty( $language ) ) {
+			// Update language.
+			if ( empty( $prepared['lang_id'] ) ) {
+				// Should not happen.
+				return new WP_Error(
+					'rest_invalid_id',
+					__( 'Invalid language ID', 'polylang' ),
+					array( 'status' => 400 )
+				);
+			}
+
+			if ( isset( $prepared['slug'], $prepared['locale'], $prepared['name'], $prepared['rtl'], $prepared['term_group'] ) ) {
+				return $prepared;
+			}
+
+			/** @phpstan-var MandatoryPropsUpdate */
+			return array_merge(
+				array(
+					'slug'       => $language->slug,
+					'locale'     => $language->locale,
+					'name'       => $language->name,
+					'rtl'        => (bool) $language->is_rtl,
+					'term_group' => $language->term_group,
+				),
+				$prepared
+			);
+		}
+
+		// Create language.
+		$error = new WP_Error(
+			'rest_invalid_locale',
+			__( 'The locale is invalid.', 'polylang' ),
+			array( 'status' => 400 )
+		);
+
+		if ( empty( $prepared['locale'] ) ) {
+			// Should not happen.
+			return $error;
+		}
+
+		if ( isset( $prepared['slug'], $prepared['name'], $prepared['rtl'], $prepared['flag'] ) ) {
+			$prepared['term_group'] ?? 0;
+			/** @phpstan-var MandatoryPropsCreate */
+			return $prepared;
+		}
+
+		$languages = include POLYLANG_DIR . '/settings/languages.php';
+
+		if ( empty( $languages[ $prepared['locale'] ] ) ) {
+			return $error;
+		}
+
+		/** @phpstan-var MandatoryPropsCreate */
+		return array_merge(
+			array(
+				'slug'       => $languages[ $prepared['locale'] ]['code'],
+				'name'       => $languages[ $prepared['locale'] ]['name'],
+				'rtl'        => 'rtl' === $languages[ $prepared['locale'] ]['dir'],
+				'flag'       => $languages[ $prepared['locale'] ]['flag'],
+				'term_group' => 0,
+			),
+			$prepared
+		);
+	}
+
+	/**
+	 * Prepares one language for create or update operation.
+	 *
+	 * @since 3.7
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return object|WP_Error The prepared language, or WP_Error object on failure.
+	 *
+	 * @phpstan-template T of LanguagePropsAll
+	 * @phpstan-param WP_REST_Request<T> $request
+	 * @phpstan-return PreparedPropsAll
+	 */
+	protected function prepare_item_for_database( $request ) {
+		$prepared = new stdClass(); // WP_REST_Controller imposes to return an object.
+		$fields   = $this->get_fields_for_response( $request );
+		$props    = array(
+			// From => To.
+			'term_id'        => 'lang_id',
+			'name'           => 'name',
+			'slug'           => 'slug',
+			'locale'         => 'locale',
+			'flag_code'      => 'flag',
+			'is_rtl'         => 'rtl',
+			'term_group'     => 'term_group',
+			'no_default_cat' => 'no_default_cat',
+		);
+
+		foreach ( $props as $rest_item_prop => $prepared_prop ) {
+			if ( isset( $request[ $rest_item_prop ] ) && rest_is_field_included( $rest_item_prop, $fields ) ) {
+				$prepared->$prepared_prop = $request[ $rest_item_prop ];
+			}
+		}
```

Comment:
This method prepares for the database, so I don't believe that we should deal with `get_fields_for response()`

Here is a proposal for the complete method (it merges what you did with `map_data_for_database()` which indeed does what I expect in this method). My understanding from the interface provided by the parent class is that this method must prepare the item for create or update. This method is thus expected to be directly called by `create()` and `update()`. 

Hopefully this proposal should help simplify all the PHPStan doc.

```suggestion
		if ( isset( $request['term_id'] ) ) {
			$language = $this->languages->get( (int) $request['term_id'] );
			
			if ( is_wp_error( $language ) ) {
				return $language;
			}
			
			$prepared->lang_id = $term_id;
			$prepared->slug    = $request['slug'] ?? $language->slug;
			$prepared->rtl     = $request['is_rtl'] ?? (bool) $language->is_rtl;	
			$prepared->flag    = $request['flag_code'] ?? $language->flag_code;	
		} else {
			// Allow to create a language from our default list with only the locale.
			if ( ! isset( $request['slug'], $request['name'], $request['is_rtl'] ) ) {
				$languages = include POLYLANG_DIR . '/settings/languages.php';

				if ( empty( $languages[ $request['locale'] ] ) ) {
					return new WP_Error(
						'pll_rest_invalid_locale',
						__( 'The locale is invalid.', 'polylang' ),
						array( 'status' => 400 )
					);
				}
					
				$language = (object) $languages[ $request['locale'] ];
			}		
			
			$prepared->slug = $request['slug'] ?? $language->code;
			$prepared->rtl  = $request['is_rtl'] ?? 'rtl' === $language->dir;
			$prepared->flag = $request['flag_code'] ?? $language->flag;
		}


		$prepared->name       = $request['name'] ?? $language->name;
		$prepared->locale     = $request['locale'] ?? $language->locale;
		$prepared->term_group = $request['term_group'] ?? $language->term_group ?? 0;

		if ( isset( request['no_default_cat'] ) ) {
			$prepared->no_default_cat = $request['no_default_cat'];
		}  
```

Reply:
It's worse. PHPStan is not smart enough.

Reply:
Done in [90cf12d](https://github.com/polylang/polylang/pull/1505/commits/90cf12d64ae9fc42a0f7a06374502b9b3041a49e).

IMHO it's not really better because PHPStan is completely failing:

1. For example, with your code as it is, there are new issues `$language may not be defined` (lines 800 to 808).
2. Doing `$args = (array) $prepared;`: `$args` looses `$prepared`'s "shape", `$args` is a classic array. Using `get_object_vars()` doesn't help.
3. The thing with `WP_REST_Request<T>` is making all this really difficult.